channels/rdpsnd/client/rdpsnd_main.c

```diff
@@ -1636,13 +1652,17 @@ static UINT rdpsnd_on_data_received(IWTSVirtualChannelCallback* pChannelCallback
 	Stream_SealLength(copy);
 	Stream_SetPosition(copy, 0);
 
-	if (!MessageQueue_Post(plugin->queue, NULL, 0, copy, NULL))
+	if (plugin->async)
 	{
-		Stream_Release(copy);
-		return ERROR_INTERNAL_ERROR;
+		if (!MessageQueue_Post(plugin->queue, NULL, 0, copy, NULL))
+		{
+			Stream_Release(copy);
+			return ERROR_INTERNAL_ERROR;
+		}
+		return CHANNEL_RC_OK;
 	}
-
-	return CHANNEL_RC_OK;
+	else
+		return rdpsnd_recv_pdu(plugin, copy);
```

Comment:
Nit-picking but imo code like this:
```
if (expr)
    return something special
return default;
```
is less error prone (i.e. to future changes that might introduce a path in the `if` or `else` that do not return) than:
```
if
    return;
else
    return
```

Reply:
Here I think it makes more sense if I just leave the last return as the last line and check error in the else path.